### PR TITLE
Begin merging NLP and NLS

### PR DESCRIPTION
**Pull request by abelsiqueira**
Part of #166 
- This is the first commit regarding merging NLP and NLS. **Does not pass tests**, because needs all updates;
- No more `AbstractNLSModel`;
- NLP is composed of single objective + NLS;
- Counters are merged.

Next parts include merging models and updating tests.

### CODE REVIEW

File: src/NLPModels.jl [hunk 1]
```diff
@@ -13,91 +12,75 @@ with `σ = obj_weight`
...
-
-"""`sum_counters(counters)`
-
-Sum all counters of `counters`.
+# Methods to be overridden in other packages.
 """
-sum_counters(c :: Counters) = sum(getfield(c, x) for x in fieldnames(Counters))
-
-"""`sum_counters(nlp)`
+    f = single_obj(nlp, x)
```
Comment (dpo) [inline src/NLPModels.jl]: What does `single_obj` represent?

File: src/NLPModels.jl [hunk 2]
```diff
@@ -13,91 +12,75 @@ with `σ = obj_weight`
...
-Reset evaluation counters
+Evaluate ``f(x)``, the objective function of `nlp` at `x`.
 """
-function reset!(counters :: Counters)
-  for f in fieldnames(Counters)
-    setfield!(counters, f, 0)
+function obj(nlp::AbstractNLPModel, x::AbstractVector)
+  increment!(nlp, :neval_obj)
+  fx = zero(eltype(x))
+  if nlp.meta.hassobj
```
Comment (dpo) [inline src/NLPModels.jl]: `has_single_obj`?

File: src/NLPModels.jl [hunk 3]
```diff
@@ -13,91 +12,75 @@ with `σ = obj_weight`
...
+    fx += nlp.σf * single_obj(nlp, x)
+  end
+  if nlp.meta.nlsequ > 0
```
Comment (dpo) [inline src/NLPModels.jl]: Should this be `nls_obj` or something like that?

File: src/NLPModels.jl (hunk 1 above)
Comment (abelsiqueira) [reply src/NLPModels.jl]: The part of the objective functions without the NLS term.

File: src/NLPModels.jl (hunk 3 above)
Comment (abelsiqueira) [reply src/NLPModels.jl]: `nlsequ` stores the number of equations in the residual. It was `nequ` before, I could change back.

Comment (dpo) [reply src/NLPModels.jl]: `nresiduals`?

Comment (dpo) [reply src/NLPModels.jl]: or `num_residuals`?

File: src/NLPModels.jl (hunk 1 above)
Comment (dpo) [reply src/NLPModels.jl]: `general_obj`?

Comment (abelsiqueira) [reply src/NLPModels.jl]: Fixed

File: src/NLPModels.jl (hunk 2 above)
Comment (abelsiqueira) [reply src/NLPModels.jl]: Fixed

File: src/NLPModels.jl (hunk 3 above)
Comment (abelsiqueira) [reply src/NLPModels.jl]: Fixed

File: src/NLPModels.jl (hunk 2 above)
Comment (dpo) [reply src/NLPModels.jl]: Should be consistent with `general_obj` I suppose. How about `has_general_obj`?

Comment (abelsiqueira) [reply src/NLPModels.jl]: Yes, I changed to that.

**Comment by abelsiqueira** [comment]
Closing this as per #199

**Pull request closed without merging**